java.lang.ClassCastException: java.io.BufferedWriter cannot be cast to java.io.PrintWriter

## Expected behavior

When doing a query on the database with valid credentials I expect the query to succeed, both when calling an endpoint that calls the database in its handler, and when calling the query from the REPL.

## Actual behavior

This suddenly stopped working (after pulling the latest CIDER, because when downgrading the error no longer appears). The query only succeeds from the REPL, when using the endpoint the following stack trace appears:

```
18-11-30 15:57:05 Erwin-Mac DEBUG [our_app.handler:62] - Looking up server: tenant.our_app.local, path: /api/cms
Could not create connection to database server.
com.mysql.jdbc.exceptions.jdbc4.MySQLNonTransientConnectionException: Could not create connection to database server.
 at sun.reflect.NativeConstructorAccessorImpl.newInstance0 (NativeConstructorAccessorImpl.java:-2)
    sun.reflect.NativeConstructorAccessorImpl.newInstance (NativeConstructorAccessorImpl.java:62)
    sun.reflect.DelegatingConstructorAccessorImpl.newInstance (DelegatingConstructorAccessorImpl.java:45)
    java.lang.reflect.Constructor.newInstance (Constructor.java:423)
    com.mysql.jdbc.Util.handleNewInstance (Util.java:404)
    com.mysql.jdbc.Util.getInstance (Util.java:387)
    com.mysql.jdbc.SQLError.createSQLException (SQLError.java:917)
    com.mysql.jdbc.SQLError.createSQLException (SQLError.java:896)
    com.mysql.jdbc.SQLError.createSQLException (SQLError.java:885)
    com.mysql.jdbc.SQLError.createSQLException (SQLError.java:860)
    com.mysql.jdbc.ConnectionImpl.connectOneTryOnly (ConnectionImpl.java:2332)
    com.mysql.jdbc.ConnectionImpl.createNewIO (ConnectionImpl.java:2085)
    com.mysql.jdbc.ConnectionImpl.<init> (ConnectionImpl.java:795)
    com.mysql.jdbc.JDBC4Connection.<init> (JDBC4Connection.java:44)
    sun.reflect.GeneratedConstructorAccessor16.newInstance (:-1)
    sun.reflect.DelegatingConstructorAccessorImpl.newInstance (DelegatingConstructorAccessorImpl.java:45)
    java.lang.reflect.Constructor.newInstance (Constructor.java:423)
    com.mysql.jdbc.Util.handleNewInstance (Util.java:404)
    com.mysql.jdbc.ConnectionImpl.getInstance (ConnectionImpl.java:400)
    com.mysql.jdbc.NonRegisteringDriver.connect (NonRegisteringDriver.java:327)
    java.sql.DriverManager.getConnection (DriverManager.java:664)
    java.sql.DriverManager.getConnection (DriverManager.java:208)
    clojure.java.jdbc$get_connection.invokeStatic (jdbc.clj:272)
    clojure.java.jdbc$get_connection.invoke (jdbc.clj:193)
    clojure.java.jdbc$db_query_with_resultset.invokeStatic (jdbc.clj:848)
    clojure.java.jdbc$db_query_with_resultset.invoke (jdbc.clj:818)
    clojure.java.jdbc$query.invokeStatic (jdbc.clj:874)
    clojure.java.jdbc$query.invoke (jdbc.clj:854)
    yesql.generate$query_handler.invokeStatic (generate.clj:89)
    yesql.generate$query_handler.invoke (generate.clj:82)
    yesql.generate$generate_query_fn$real_fn__32447.invoke (generate.clj:120)
    yesql.generate$generate_query_fn$query_wrapper_fn__32454.invoke (generate.clj:142)
    clojure.lang.AFn.applyToHelper (AFn.java:156)
    clojure.lang.AFn.applyTo (AFn.java:144)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:29)
    clojure.lang.RestFn.invoke (RestFn.java:421)
    our_app.handler$extract_tenant_id_from_subdomain.invokeStatic (handler.clj:63)
    our_app.handler$extract_tenant_id_from_subdomain.invoke (handler.clj:57)
    our_app.handler$wrap_tenant_id$fn__46357.invoke (handler.clj:96)
    ring.middleware.cookies$wrap_cookies$fn__22430.invoke (cookies.clj:175)
    ring.middleware.session$wrap_session$fn__22692.invoke (session.clj:108)
    ring.middleware.reload$wrap_reload$fn__23609.invoke (reload.clj:39)
    our_app.handler$wrap_echo_url$fn__46383.invoke (handler.clj:233)
    clojure.lang.Var.invoke (Var.java:381)
    figwheel_sidecar.components.figwheel_server$possible_endpoint$fn__68452.invoke (figwheel_server.clj:224)
    figwheel_sidecar.components.figwheel_server$handle_static_resources$fn__68432.invoke (figwheel_server.clj:197)
    figwheel_sidecar.components.figwheel_server$handle_index$fn__68423.invoke (figwheel_server.clj:183)
    figwheel_sidecar.components.figwheel_server$handle_figwheel_websocket$fn__68446.invoke (figwheel_server.clj:219)
    figwheel_sidecar.components.figwheel_server$wrap_no_cache$fn__68457.invoke (figwheel_server.clj:239)
    co.deps.ring_etag_middleware$wrap_file_etag$fn__67808.invoke (ring_etag_middleware.clj:94)
    ring.middleware.not_modified$wrap_not_modified$fn__21529.invoke (not_modified.clj:53)
    ring.middleware.cors$handle_cors.invokeStatic (cors.clj:170)
    ring.middleware.cors$handle_cors.invoke (cors.clj:159)
    ring.middleware.cors$wrap_cors$fn__22497.invoke (cors.clj:184)
    figwheel_sidecar.components.figwheel_server$log_output_to_figwheel_server_log$fn__68414.invoke (figwheel_server.clj:143)
    org.httpkit.server.HttpHandler.run (RingHandler.java:91)
    java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    java.util.concurrent.FutureTask.run (FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    java.lang.Thread.run (Thread.java:748)

Caused by: java.lang.ClassCastException: java.io.BufferedWriter cannot be cast to java.io.PrintWriter
 at cider.nrepl.middleware.out$print_stream$fn__53158.invoke (out.clj:99)
    cider.nrepl.middleware.out.proxy$java.io.OutputStream$ff19274a.write (:-1)
    java.io.PrintStream.write (PrintStream.java:480)
    sun.nio.cs.StreamEncoder.writeBytes (StreamEncoder.java:221)
    sun.nio.cs.StreamEncoder.implFlushBuffer (StreamEncoder.java:291)
    sun.nio.cs.StreamEncoder.flushBuffer (StreamEncoder.java:104)
    java.io.OutputStreamWriter.flushBuffer (OutputStreamWriter.java:185)
    java.io.PrintStream.write (PrintStream.java:527)
    java.io.PrintStream.print (PrintStream.java:669)
    java.io.PrintStream.println (PrintStream.java:806)
    com.mysql.jdbc.log.StandardLogger.logInternal (StandardLogger.java:318)
    com.mysql.jdbc.log.StandardLogger.logWarn (StandardLogger.java:240)
    com.mysql.jdbc.MysqlIO.doHandshake (MysqlIO.java:1177)
    com.mysql.jdbc.ConnectionImpl.coreConnect (ConnectionImpl.java:2255)
    com.mysql.jdbc.ConnectionImpl.connectOneTryOnly (ConnectionImpl.java:2286)
    com.mysql.jdbc.ConnectionImpl.createNewIO (ConnectionImpl.java:2085)
    com.mysql.jdbc.ConnectionImpl.<init> (ConnectionImpl.java:795)
    com.mysql.jdbc.JDBC4Connection.<init> (JDBC4Connection.java:44)
    sun.reflect.GeneratedConstructorAccessor16.newInstance (:-1)
    sun.reflect.DelegatingConstructorAccessorImpl.newInstance (DelegatingConstructorAccessorImpl.java:45)
    java.lang.reflect.Constructor.newInstance (Constructor.java:423)
    com.mysql.jdbc.Util.handleNewInstance (Util.java:404)
    com.mysql.jdbc.ConnectionImpl.getInstance (ConnectionImpl.java:400)
    com.mysql.jdbc.NonRegisteringDriver.connect (NonRegisteringDriver.java:327)
    java.sql.DriverManager.getConnection (DriverManager.java:664)
    java.sql.DriverManager.getConnection (DriverManager.java:208)
    clojure.java.jdbc$get_connection.invokeStatic (jdbc.clj:272)
    clojure.java.jdbc$get_connection.invoke (jdbc.clj:193)
    clojure.java.jdbc$db_query_with_resultset.invokeStatic (jdbc.clj:848)
    clojure.java.jdbc$db_query_with_resultset.invoke (jdbc.clj:818)
    clojure.java.jdbc$query.invokeStatic (jdbc.clj:874)
    clojure.java.jdbc$query.invoke (jdbc.clj:854)
    yesql.generate$query_handler.invokeStatic (generate.clj:89)
    yesql.generate$query_handler.invoke (generate.clj:82)
    yesql.generate$generate_query_fn$real_fn__32447.invoke (generate.clj:120)
    yesql.generate$generate_query_fn$query_wrapper_fn__32454.invoke (generate.clj:142)
    clojure.lang.AFn.applyToHelper (AFn.java:156)
    clojure.lang.AFn.applyTo (AFn.java:144)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:29)
    clojure.lang.RestFn.invoke (RestFn.java:421)
    our_app.handler$extract_tenant_id_from_subdomain.invokeStatic (handler.clj:63)
    our_app.handler$extract_tenant_id_from_subdomain.invoke (handler.clj:57)
    our_app.handler$wrap_tenant_id$fn__46357.invoke (handler.clj:96)
    ring.middleware.cookies$wrap_cookies$fn__22430.invoke (cookies.clj:175)
    ring.middleware.session$wrap_session$fn__22692.invoke (session.clj:108)
    ring.middleware.reload$wrap_reload$fn__23609.invoke (reload.clj:39)
    our_app.handler$wrap_echo_url$fn__46383.invoke (handler.clj:233)
    clojure.lang.Var.invoke (Var.java:381)
    figwheel_sidecar.components.figwheel_server$possible_endpoint$fn__68452.invoke (figwheel_server.clj:224)
    figwheel_sidecar.components.figwheel_server$handle_static_resources$fn__68432.invoke (figwheel_server.clj:197)
    figwheel_sidecar.components.figwheel_server$handle_index$fn__68423.invoke (figwheel_server.clj:183)
    figwheel_sidecar.components.figwheel_server$handle_figwheel_websocket$fn__68446.invoke (figwheel_server.clj:219)
    figwheel_sidecar.components.figwheel_server$wrap_no_cache$fn__68457.invoke (figwheel_server.clj:239)
    co.deps.ring_etag_middleware$wrap_file_etag$fn__67808.invoke (ring_etag_middleware.clj:94)
    ring.middleware.not_modified$wrap_not_modified$fn__21529.invoke (not_modified.clj:53)
    ring.middleware.cors$handle_cors.invokeStatic (cors.clj:170)
    ring.middleware.cors$handle_cors.invoke (cors.clj:159)
    ring.middleware.cors$wrap_cors$fn__22497.invoke (cors.clj:184)
    figwheel_sidecar.components.figwheel_server$log_output_to_figwheel_server_log$fn__68414.invoke (figwheel_server.clj:143)
    org.httpkit.server.HttpHandler.run (RingHandler.java:91)
    java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    java.util.concurrent.FutureTask.run (FutureTask.java:266)
    java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    java.lang.Thread.run (Thread.java:748)
```

## Steps to reproduce the problem

I don't see how to make the problem smaller than it is. We are calling a database via yesql generated queries from a handler and it works from the REPL, but fails when we call the endpoint that calls the handler. When running `lein figwheel` and `lein repl` from a shell everything works fine. Also when downgrading (by manually setting the `cider-latest-middleware-version` in `cider.el` to 0.18.0 and recompiling) it works again.

In the stack trace we see:

`Caused by: java.lang.ClassCastException: java.io.BufferedWriter cannot be cast to java.io.PrintWriter
 at cider.nrepl.middleware.out$print_stream$fn__53158.invoke (out.clj:99)`

I don't know why a BufferedWriter is in the `@printer` variable, but what I do know is that the latest commit https://github.com/clojure-emacs/cider-nrepl/commit/658194a8abb0842cbc3efe4f7cb04917487785d2 introduces the type hints that `@printer` should be a `PrintWriter`. In this case `@printer` is a `BufferedWriter`. Both classes inherit from [`Writer`](https://docs.oracle.com/javase/7/docs/api/java/io/Writer.html) and only write is called so I expect the code to work again when either changing the type hint to `Writer` or removing it.

## Environment & Version information

MySQL 5.7.19 Homebrew
yesql 0.5.3
Spacemacs 0.200.13@26.1
Emacs 26.1

### cider-nrepl version

`0.19.0-SNAPSHOT` (`cider-20181128.1311`, after above mentioned commit)

### Java version

1.8.0_192

### Operating system

macOS High Sierra

--- a/src/cider/nrepl/middleware/out.clj
+++ b/src/cider/nrepl/middleware/out.clj
@@ -81,7 +81,7 @@
         print-timer (Timer.)
         print-flusher (proxy [TimerTask] []
                         (run []
-                          (.flush ^PrintWriter @printer)))]
+                          (.flush ^Writer @printer)))]
     (.scheduleAtFixedRate print-timer print-flusher delay delay)
     (PrintStream.
      (proxy [OutputStream] []
@@ -92,13 +92,13 @@
        (write
          ([int-or-bytes]
           (if (instance? Integer int-or-bytes)
-            (.write ^PrintWriter @printer ^Integer int-or-bytes)
-            (.write ^PrintWriter @printer (String. ^"[B" int-or-bytes))))
+            (.write ^Writer @printer ^Integer int-or-bytes)
+            (.write ^Writer @printer (String. ^"[B" int-or-bytes))))
          ([^"[B" bytes ^Integer off ^Integer len]
           (let [byte-range (byte-array (take len (drop off bytes)))]
-            (.write ^PrintWriter @printer (String. byte-range)))))
+            (.write ^Writer @printer (String. byte-range)))))
        (flush []
-         (.flush ^PrintWriter @printer))))))
+         (.flush ^Writer @printer))))))
 
 ;;; Known eval sessions
 (def tracked-sessions-map

